src/devices/efergy_e2_classic.c

```diff
@@ -60,24 +68,59 @@ static int efergy_e2_classic_callback(bitbuffer_t *bitbuffer) {
 		return 0;
 	}
 
-	const unsigned VOLTAGES[] = {110, 115, 120, 220, 230, 240, 0};
-
-	double current_adc = 256 * bytes[4] + bytes[5];
-	for (unsigned i = 0; VOLTAGES[i] != 0; ++i) {
-		double power  = (VOLTAGES[i] * current_adc * (1 << bytes[6])) / 32768;
-		fprintf(stderr, "Power consumption at %u volts: %.2f watts\n", VOLTAGES[i], power);
+	if (bytes[6] > 0x80)
```

Comment:
Good thing you caught that. That looked clean but was entirely stupid of me. Sorry. Convert to float (it will be double internally, there should be no floats in the code base… discussion for another day, though) apply scaling factor. Just like you originally had.

Dividing by 0x8000 is essentially a shift-right by 15. This results in a factor of pow(2, fact-15). I take it this is because the "mantissa" is 16-bit and needs to be normalized to 0<M<1. I think I get it now!

There is integer promotion, arithmetic conversion, and the cast operator in C. The cast `(int8_t)` will reinterpret here (and the result will actually be `int` …integer promotion with sign extension!). Other places like `(float) someint` actually trigger arithmetic conversion. C is misleading at best :)

Also, your `pow()` call will promote the signed char to integer and then convert to double. The const `2` will be converted to double at compile time. To skip some of this, it's best to declare temporary variables, like `fact` as integer -- promotion isn't that expensive but will likely lead to wrong assumptions.